hw-model/src/lib.rs

```diff
@@ -137,6 +137,36 @@ pub trait HwModel {
         }
     }
 
+    fn copy_output_until_fatal_error(
```

Comment:
The semantics of this function confuse me. My expectation was that it would only succeed if the firmware failed with an error code matching the argument, and fail if the firmware did an "exit success" or "exit with some other failure".

If you want the semantics as-written, perhaps it makes more sense to add the error-checking logic to copy_output_until_exit_success?

Reply:
Updated to only succeed if the error code matches.

Reply:
> If you want the semantics as-written, perhaps it makes more sense to add the error-checking logic to copy_output_until_exit_success?

Not a huge deal, but this suggestion seems more clear rather than having different functions for checking different kinds of errors

Reply:
Let's go with the current change; we'll revisit if this ends up into too many copy_output_* functions.